version: '3.8'

services:
  web:
    build: .
    expose:
      - "5001"
    depends_on:
      - db
    environment:
      - FLASK_APP=app/main.py
      - FLASK_ENV=development
      - FLASK_DEBUG=1
      - DATABASE_URL=postgresql://postgres:password@db:5432/flask_db
      - OPENAI_API_KEY=${OPENAI_API_KEY}
    volumes:
      - .:/app
      - ./app/static/uploads:/app/app/static/uploads
    networks:
      - diary-network
    command: flask run --host=0.0.0.0 --port=5001 --debug

  nginx:
    image: nginx:latest
    ports:
      - "80:80"
    volumes:
      - ./nginx/conf:/etc/nginx/conf.d
      - .:/app
      - ./nginx/certbot/www:/var/www/certbot
    depends_on:
      - web
    networks:
      - diary-network

  certbot:
    image: certbot/certbot
    volumes:
      - ./nginx/certbot/conf:/etc/letsencrypt
      - ./nginx/certbot/www:/var/www/certbot
    depends_on:
      - nginx
    entrypoint: "/bin/sh -c 'trap exit TERM; while :; do certbot renew; sleep 12h & wait $${!}; done;'"

  db:
    image: postgres:13
    volumes:
      - postgres_data:/var/lib/postgresql/data
    environment:
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=password
      - POSTGRES_DB=flask_db
    ports:
      - "5432:5432"
    networks:
      - diary-network

volumes:
  postgres_data:

networks:
  diary-network:
    driver: bridge 